pgjdbc/src/main/java/org/postgresql/jdbc/PgDatabaseMetaData.java

```diff
@@ -2224,7 +2224,9 @@ public ResultSet getTypeInfo() throws SQLException {
     String sql;
     sql = "SELECT t.typname,t.oid FROM pg_catalog.pg_type t"
           + " JOIN pg_catalog.pg_namespace n ON (t.typnamespace = n.oid) "
-          + " WHERE n.nspname  != 'pg_toast'";
+          + " WHERE n.nspname  != 'pg_toast'"
+          + " AND "
+          + " (t.typrelid = 0 OR (SELECT c.relkind = 'c' FROM pg_catalog.pg_class c WHERE c.oid = t.typrelid))";
```

Comment:
What's `c.relkind = 'c'`? I don't see it in the list of values for relkind. Do you mean `'r'` for ordinary tables?

If we're going to go the route of filtering out table types why not remove the relkind check entirely? That way anything that is a "thing with columns" would get filtered out too. So that means, views, mviews, partitioned tables, foreign tables, and anything else new that comes along.

Reply:
Ah my mistake. I read the logic backwards. Yes that makes sense.

I doubt it makes any perf difference in the real world but the subquery could be split out to left join to `pg_class`. Maybe a bit easier to understand too:

```sql
SELECT
  t.typname,
  t.oid
FROM pg_catalog.pg_type t
  JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace
  LEFT JOIN pg_class c ON c.oid = t.typrelid
WHERE n.nspname != 'toast'
  AND (t.typrelid = 0 OR c.relkind = 'c')
```

Reply:
I borrowed it from psql -E 
/dT

:)